UPE - Support free trial subscriptions

ricardo opened the pull request:
Fixes #1867

# Changes proposed in this Pull Request:

- If order is being processed with a setup intent and doesn't require payment, complete the order right away, like `process_payment` for credit card payments.
- Use a setup intent for displaying the UPE payment fields in case the product doesn't require payment.

# Testing instructions

- Ensure you have the UPE feature flag enabled.
- Install [WooCommerce Subscriptions](https://woocommerce.com/products/woocommerce-subscriptions/), in case you haven't already.
- Navigate to **Products -> Add new**
- Under "Product data", select simple subscription, and add a subscription price and a free trial period.
- As a shopper, add the newly created free trial subscription product to the cart and try to complete the checkout.
- Ensure the checkout works and the order is moved to "processing" or "completed".
- Ensure the subscription is made active.
- As a merchant, navigate to "WooCommerce -> Subscriptions" and open the newly created subscription.
- Ensure you can process a renewal under "Subscription actions".
- Ensure a new order is attached to the subscription under "Related Orders", and the status of that order is "processing" or "completed".

ricardo committed:
Process setup intents

ricardo committed:
Pass `isPaymentRequired` param to front-end to decide whether or not a setup intent should be used

diegocurbelo reviewed (approved):
Tests well and code looks good! 🚢 

I didn't add the `ready to merge` because the base is `fix/1665-upe-subscriptions-support`

ricardo committed:
Free trial subs compatible with checkout block

ricardo commented:
@diegocurbelo I rebased and added compatibility for the block checkout as well in 01dc9a18f7254f3d986aaa5ad9d3eeeedebab60d.

Can you give this another look and let me know if it's good to merge?

diegocurbelo reviewed (commented):
I'm not sure if this is a problem of this PR or something that we should fix in another PR, but...

- When using the Shortcode checkout it all works as expected.. in the subscription details the payment method is `Credit Card (Stripe)` and the renewal completes.

- But when using the Blocks checkout the subscription payment method is: `Manual Renewal`, so after a renewal the status is `Pending payment`


Other than that, the code looks good, and with the shortcode checkout it tests well

ricardo committed:
Avoid payment intent error in Stripe by skipping payment intent update if it's a setup intent id

ricardo commented:
@diegocurbelo Nice find! 👍 

I'm suspecting this is an issue with subscriptions + the blocks checkout endpoint. It works in the classic checkout probably because it doesn't use the blocks API. This issue also happens when UPE is disabled.

For some reason, when the order contains a free subscription, it sets the `_requires_manual_renewal` post meta to `true`, and doesn't set the payment method for the subscription. That's not controlled on our end, so I'm guessing the issue isn't on us. - I'll need more time to investigate that, but for the time being I believe we can merge as is, since this wasn't working before and we don't seem to be breaking anything.

I noticed there was a Stripe error log and added a small change in 5b194eb43c54f08a71e75d9ee22b8a3fd8827319. I'm not sure why the intent is being updated every time the order is complete (perhaps @reykjalin can answer that?), but it shouldn't update setup intents.

diegocurbelo reviewed (approved):
Looks good to merge 🚢

ricardo committed:
Merge branch 'develop' into fix/1867-free-trial-subscriptions

reykjalin commented:
> I'm not sure why the intent is being updated every time the order is complete (perhaps @reykjalin can answer that?), but it shouldn't update setup intents.

I don't think I implemented any separate flows for setup and payment intents in the Blocks integration. Maybe that's the root cause? 🤔
